Arch arm mpu move internal api

ioannisg opened the pull request:
A clean-up patch; moving an internal API header where it belongs (IMHO) :)

ioannisg committed:
arch: arm: mpu: move internal API inside arch/arm/cortex_m/mpu

arm_core_mpu_dev.h is an internal API, and is not supposed to
be directly called by kernel / application functions, therefore,
we can move it inside arch/arm/core/cortex_m/mpu directory.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
tests: kernel: userspace: explicitly declare function for the test

In the wake of moving the internal API header arm_core_mpu_dev.h
into arch/arm/cortex_m/mpu, we need to explicitly declare the
arm_core_mpu_disable() function in the userspace test. Note that
arm_core_mpu_disable() (as any other function in this internal
API) is not supposed to be called directly by kernel/application
functions; an exception is allowed in this test suite, so we are
able to test the MPU disabling functionality.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

agansari reviewed (approved):
@ioannisg looks okay

galak commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -9,7 +9,7 @@
 #include <soc.h>
 
-#include <arch/arm/cortex_m/mpu/arm_core_mpu_dev.h>
+#include <arm_core_mpu_dev.h>
```
can this be "arm_core_mpu_dev.h" ?

ioannisg replied:
Yeap, changed this now.

galak commented on tests/kernel/mem_protect/userspace/src/main.c:
```diff
@@ -23,7 +23,7 @@
 
 #if defined(CONFIG_ARM)
-#include <arch/arm/cortex_m/mpu/arm_core_mpu_dev.h>
+extern void arm_core_mpu_disable(void);
```
So this made me look at the actual test, why do we have a disable_mmu_mpu() if we don't have an exposed API related to it?

ioannisg replied:
@andrewboie FYI

ioannisg replied:
@galak I get the feeling that this test is there to show that we can't disable the MPU from user mode, right @andrewboie ?

andrewboie replied:
I wouldn't object if we deleted that test. There's some other dubious cases in there as well.

ioannisg replied:
I like that it "tests" that user-mode can't play with the mpu :)

andrewboie replied:
Are we trying to prove whether the CPU works or the software?
Proving that the CPU works is a waste of time IMO

ioannisg replied:
Yeap, that's in fact, correct

galak reviewed (commented):
Changes looks fine, but curious about why arm_core_mpu_disable has to get exported and what actually the test case is doing in disable_mmu_mpu().

Feels odd to me that we are "Testing" something that doesnt have an exposed API.

galak reviewed (dismissed):
Can we remove arm_core_mpu_disable() from arm_core_mpu_dev.h since its effectively static and the only non-static use is the explicit extern in the testcase.

ioannisg commented:
> 
> 
> Can we remove arm_core_mpu_disable() from arm_core_mpu_dev.h since its effectively static and the only non-static use is the explicit extern in the testcase.

Hm, I think yes - we used to invoke the enable/disable functions in arm_core_mpu.c, but perhaps not anymore.

ioannisg committed:
arch: arm: mpu: remove arm_core_mpu_enable/disable internal API

arm_core_mpu_enable() and arm_core_mpu_disable() functions are
effectively static functions, only used in the drivers for ARM
and NXP MPU, therefore, we do not need to expose them in the
arm_core_mpu_dev.h header.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
> 
> 
> Can we remove arm_core_mpu_disable() from arm_core_mpu_dev.h since its effectively static and the only non-static use is the explicit extern in the testcase.

Done now, thanks.

galak commented:
LGTM, will wait to see if any discussion from @andrewboie on what the test is trying to do.
